### PR TITLE
Show approvals on the PR listing page

### PR DESCRIPTION
**Pull request by jamesorlakin**
As suggested in #8659, this PR displays the whitelisted approvals in Gitea's PR listing page if a given branch is protected and requires more than 0 approvals to be merged. I've chosen the eye Octicon but perhaps a person one would be better?

image

I've tested this works locally in the browser but as I'm on a Windows environment Docker doesn't work properly for the full test suite 😢 . This is also my first time using Go, so all suggestions (nitpicking) are most welcome!

### CODE REVIEW

File: models/branches.go
Comment (lafriks) [inline models/branches.go]: This should not be added column so database so should be marked so:
```suggestion
	GrantedApprovalsCount     int64    `xorm:"-"`
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/branches.go
Comment (6543) [reply models/branches.go]: granted aprovals Should be stored for each PR and not for each branch?!

Comment (6543) [reply models/branches.go]: And the information Iis stored already somewhere ... have to check where ...

File: templates/repo/issue/list.tmpl
Comment (6543) [inline templates/repo/issue/list.tmpl]:
```suggestion
								<span class="comment ui right"><i class="octicon octicon-eye"></i> {{.PullRequest.GrantedApprovalsCount}} / {{.PullRequest.ProtectedBranch.RequiredApprovals}}</span>
```
Continue things .. from abouth

**Review by 6543 — CHANGES_REQUESTED** [review]
Beside this ... nice idear

File: models/branches.go
Comment (jamesorlakin) [reply models/branches.go]: Aah, so any properties I add are mapped to the database _by default_?

File: templates/repo/issue/list.tmpl
Comment (jamesorlakin) [reply templates/repo/issue/list.tmpl:225]: Aah that's much simpler, will do

File: models/branches.go
Comment (6543) [reply models/branches.go]: you can get the information via `models.GetUniqueApprovalsByPullRequestID(prID int64) ([]*Review, error)`

Comment (6543) [reply models/branches.go]: and for the template engine you should provide them via `routers/repo/issue.go:`
as `IsBlockedByApprovals` value

Comment (6543) [reply models/branches.go]: summary: 
* no changes in models/branches.go
* call models.GetUniqueApprovalsByPullRequestID(prID int64) in routers/repo/issue.go and store in `ctx.Data["GrantedApprovalsCount"]`

**Comment by 6543** [comment]
@jamesorlakin look at https://github.com/go-gitea/gitea/pull/8762#discussion_r341155966 ;)

File: models/branches.go
Comment (jamesorlakin) [reply models/branches.go]: > call models.GetUniqueApprovalsByPullRequestID(prID int64) in routers/repo/issue.go and store in ctx.Data["GrantedApprovalsCount"]

How would I reference that per PR, seeing as I have an array of them? Issues get popped under `ctx.Data["Issues"]` as a whole so I created a property to be able to access it from each one within the template. I presume something dynamic like `ctx.Data["Issues"][i].GrantedApprovals = count` isn't possible - I come from a JS background 😉.

The reason I left it as a property of `ProtectedBranch` was simply because `GetGrantedApprovalsCount` lived there already, and I just assumed I would keep them together.

Comment (6543) [reply models/branches.go]: no

Comment (6543) [reply models/branches.go]: I can look into it later ...

**Comment by lunny** [comment]
I think you guys should consider the page load performance. Protected branches could be load once. Or I should send another PR to do that after this one.

File: routers/repo/issue.go
Comment (kolaente) [inline routers/repo/issue.go]: You're not handling the error here. Please do something like
```
if err := pull.LoadProtectedBranch(); err != nil {
    return
}

if pull.ProtectedBranch != nil && pull.ProtectedBranch.RequiredApprovals != 0 {
    ProtectedBranch.GrantedApprovalsCount = pull.ProtectedBranch.GetGrantedApprovalsCount(
}
```

**Review by kolaente — CHANGES_REQUESTED** [review]

File: models/branches.go
Comment (jamesorlakin) [reply models/branches.go]: I'm attempting to change it to a solution similar to `ctx.Data["CommitStatus"]` and removing the extra property on `ProtectedBranch`.

Looking at it, `GetUniqueApprovalsByPullRequestID` will return all approved reviews for a given PR, not just those on the approval whitelist. This might be equally valid - what would the user reasonably expect to see? Personally I think only reviews "which matter" should count to quickly determine if a PR is ready to be merged.

File: routers/repo/issue.go
Comment (jamesorlakin) [reply routers/repo/issue.go]: Aah ok, is there a subtle difference in the behaviour of errors and nil?

File: templates/repo/issue/list.tmpl
Comment (jamesorlakin) [reply templates/repo/issue/list.tmpl:225]: Unfortunately this didn't work for branches without protection enabled

**Comment by jamesorlakin** [comment]
Hi all,

Thanks for the comments. I've reworked it to now use a method similar to the last commit status - a map by PR ID for both required approvals and approved accepted ones.

6543 mentioned using `GetUniqueApprovalsByPullRequestID`. This would surely count all reviews on a PR, not just ones whitelisted needed for the PR to be merged? This raised a good point in my mind though - should the user see _all_ approvals, or just ones relevant to the branch protection?

lunny, you're right, I didn't think about performance! 3 PRs rendered quick enough on my laptop... A to-do (assuming this approach looks okay) would be to cache protected branches during the loop - say a page had 25 PRs all for the same target branch, we would only need to load the `ProtectedBranch` once leaving just a call to `GetGrantedApprovalsCount` for each one.

**Comment by guillep2k** [comment]
The number of protected branches in a given repository should not be too large (3? 6? *10*?). In my experience, it's more performant to load all of them at once rather than reaching for them one by one. Databases do that themselves: for example, for tables under a certain number of rows, indexes are never used for querying (only for updating); the database considers that reading one disk block is the "expensive" part; sorting/filtering is quickly done in memory. Context/process switching from Gitea to and from the database is costly as well.

**Comment by guillep2k** [comment]
For the number of approvals, I'd use a single database query, which is much faster than retrieving the information each time:

* Collect all the issues/prs to show in the page using `models.Issues()` as normal, so pagination and other options are taken into account.
* In a slice (array), collect all the ids for rows where `.IsPull` is true and `.IsClosed` is false.
* If the number of ids collected is not zero, make one single query to the database counting approvals for those ids (using group by and count); load that into a map. For that, a method must be created in `models/issue.go`, where all database accesses should happen. That method should use `sess.GroupBy("issue.id").Select("issue.id AS id, COUNT(*) AS count")`; see `CountIssuesByRepo()` for an example.
* Then that map can be used to display the info in the page.
* Alternatively, the approval count can be moved into `models.Issues()` as well.

Hope this helps.

File: routers/repo/issue.go
Comment (kolaente) [reply routers/repo/issue.go]: If the function returns `nil` for an error, there was no error. You don't check for the case where the error was not `nil` which means you're never actually doing something with the error.

In Go, there is this pattern of keeping a "happy path" of code, that's why you'll see a lot of `if err != nil ` code used to handle errors. Take a look at this article for more information: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by jamesorlakin** [comment]
Thanks for the suggestions @guillep2k! Unfortunately I'm not sure I'm capable at dropping to such complex SQL (joining and comparing whitelisted approvers including teams, etc). That said, loading all protected branches at once initially could be better optimised than the individual looping - I'll try and do that.

> ids for rows where .IsPull is true and .IsClosed is false.

I would be under the assumption closed PRs should show the same as open ones?

Moving forward, I'm also considering showing total approvals (not whitelisted) for PRs that aren't on a protected branch. This would just be a simple number next to the eye. If possible I'll want to grab all approvals for the PRs at once, and then do the whitelist filtering using a util function (similar to how `GetGrantedApprovalsCount` works, but without the query) - this may not go to plan...

I've since backported this to our production 1.9 instance and performance seems to be fine - though it's not a great representative measurement!

**Comment by 6543** [comment]
@guillep2k I can creat SQL statements but i dont know how to tell xorm to do the same :(
-> think @jamesorlakin need some help ...

**Comment by 6543** [comment]
@jamesorlakin do you still working on this?

**Comment by jamesorlakin** [comment]
@6543 yep I'm keen to finish it, but I likely won't have much time until the weekend. Feel free to tweak if you want to in the meantime 😃 .

File: templates/repo/issue/list.tmpl
Comment (davidsvantesson) [reply templates/repo/issue/list.tmpl:225]: I suppose that requires that $.RequiredApprovals contain all indexes? Should be possible by setting them 0 in issue.go.

File: routers/repo/issue.go
Comment (davidsvantesson) [inline routers/repo/issue.go:246]: Set value in slice also when branch is not protected? (with 0 value)

**Comment by 6543** [comment]
@jamesorlakin what is the stats?

**Comment by 6543** [comment]
@jamesorlakin ping

**Comment by jamesorlakin** [comment]
Hi all,

Apologies for the general radio silence - everything has been a tad busy of recent and it all generally got forgotten about 😥. In the meantime I've had a look around and #9274 seems to be a generally more polished implementation if you would like to go ahead with that.

**Pull request closed without merging**
